stackslib/src/chainstate/nakamoto/coordinator/tests.rs

```diff
@@ -101,21 +109,118 @@ fn advance_to_nakamoto(
                     )
                 })
                 .collect()
+        } else if sortition_height == 8 {
+            with_sortdb(peer, |chainstate, sortdb| {
+                make_all_signers_vote_for_aggregate_key(
+                    chainstate,
+                    sortdb,
+                    &tip.unwrap(),
+                    test_signers,
+                    test_stackers,
+                    7,
+                )
+            })
         } else {
             vec![]
         };
 
-        peer.tenure_with_txs(&txs, &mut peer_nonce);
+        tip = Some(peer.tenure_with_txs(&txs, &mut peer_nonce));
     }
     // peer is at the start of cycle 8
 }
 
+pub fn make_all_signers_vote_for_aggregate_key(
+    chainstate: &mut StacksChainState,
+    sortdb: &SortitionDB,
+    tip: &StacksBlockId,
+    test_signers: &mut TestSigners,
+    test_stackers: &[TestStacker],
+    cycle_id: u128,
+) -> Vec<StacksTransaction> {
+    info!("Trigger signers vote for cycle {}", cycle_id);
+
+    // Check if we already have an aggregate key for this cycle
+    if chainstate
+        .get_aggregate_public_key_pox_4(sortdb, tip, cycle_id as u64)
+        .unwrap()
+        .is_some()
+    {
+        debug!("Aggregate key already set for cycle {}", cycle_id);
+        return vec![];
+    }
+
+    // Generate a new aggregate key
+    test_signers.generate_aggregate_key(cycle_id as u64);
+
+    let signers_res = readonly_call_with_sortdb(
+        chainstate,
+        sortdb,
+        tip,
+        SIGNERS_NAME.into(),
+        "get-signers".into(),
+        vec![Value::UInt(cycle_id)],
+    );
+
+    // If the signers are not set yet, then we're not ready to vote yet.
+    let signer_vec = match signers_res.expect_optional().unwrap() {
+        Some(signer_vec) => signer_vec.expect_list().unwrap(),
+        None => {
+            debug!("No signers set for cycle {}", cycle_id);
+            return vec![];
+        }
+    };
+
+    let mut signers_to_index = HashMap::new();
+    for (index, value) in signer_vec.into_iter().enumerate() {
```

Comment:
What if there are duplicate signers?  I think the code that calls `set-signers` guarantees this, but perhaps there should be an assertion here as well?

Reply:
Added in 51f0c9142.